docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,602 @@
+```ruby
+class User::Policy::CanUpdateUsername < Service::PolicyBase
+  delegate :user, to: :context, private: true
+
+  def call
+    guardian.can_edit_username?(user)
+  end
+
+  def reason
+    # Here we can put more complex logic to dynamically output a reason, this is just an example
+    "Can't edit username"
+  end
+end
+```
+
+There are some rules to keep in mind when writing a policy object:
+- it has to inherit from `Service::PolicyBase`.
+- it has to define two methods: `#call` and `#reason`.
+- the context object is automatically injected in the policy, and is available by calling `#context` (like in a service).
+- likewise, the guardian object is also automatically available as `#guardian`.
+- by convention, it should be namespaced under its concept followed by the `Policy` namespace: for our current example, it means `User::Policy::` which maps to `app/services/user/policy/` on the filesystem.
+
+> 💡 To keep things short and clear, feel free to use [`delegate`](https://api.rubyonrails.org/classes/Module.html#method-i-delegate) extensively.
+
+Then, when you want to use it in a service, just write your step like this:
+```ruby
+policy :can_update_username, class_name: User::Policy::CanUpdateUsername
+```
+
+## Actions
+
+When a step starts becoming too complex, like it has too many branching statements for example, then it’s time to extract all that logic to a dedicated class. That logic could live in a model for instance, but when in doubt, just create a new action.
+An action is just a small class that responds to a `.call` method by convention. What happens inside is up to you. The idea, however, is to execute an action (hence the name) with minimal overhead. It means an action should not validate data, for example. It should be called with valid objects only, thus being able to work with them right away. It also means that an action should not fail. You can think of an action as a bare-bones service without all the bells and whistles. Also, an action can be reused by different services.
+
+Here again, it’s relatively simple to create a new action. Let’s take as an example our `log` step:
+```ruby
+class User::Action::LogUsernameChange < Service::ActionBase
+  option :actor
+  option :user
+
+  def call
+    StaffActionLogger.new(actor).log_username_change(
+      user,
+      user.username_before_last_save,
+      user.username,
+    )
+  end
+end
+```
+Of course, this is a very basic example, you can do more complex things in an action. A real-world example can be found in [`User::Action::TriggerPostAction`](https://github.com/discourse/discourse/blob/main/app/services/user/action/trigger_post_action.rb).
+`Service::ActionBase` comes with [`Dry::Initializer`](https://dry-rb.org/gems/dry-initializer) which provides a nice mini-DSL:
+- use `param :my_arg` to declare a required argument named `my_arg`.
+- use `option :my_arg` to declare a required named argument named `my_arg`.
```

Comment:
That’s the DSL from `dry-initializer`, `param` for “standard” parameters and `option` for keyword ones. But maybe that’s confusing, and it’s simpler to just write about `option` as that would be the preferred way anyway.

Reply:
I tried to make it clearer by using the proper terms (positional/keyword arguments), tell me if that’s better :sweat_smile:

Reply:
I feel like it would make more sense to only tell people about `option` since it's the preferred way and keyword arguments feel like they would be better here anyway

Reply:
Done!